### PR TITLE
Hadoop on lustre2

### PR DESCRIPTION
**Pull request by ehiggs**
Configs that support Hadoop-on-lustre2 and Hadoop-2.5.0-cdh5.3.1 using GPFS.

These versions of Hadoop require the native libraries to be built and an EasyBuild block/config is in the works.

### CODE REVIEW

File: etc/hod/Hadoop-2.5.0-cdh5.3.1-gpfs/hod.conf [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+#workdir=
+directories=$localworkdir/dfs/name,$localworkdir/dfs/data
+
+[core-site.xml]
+hadoop.tmp.dir=$localworkdir/tmp
+dfs.replication=1
+fs.defaultFS=gpfs:///
+fs.gpfs.impl=org.apache.hadoop.fs.gpfs.GeneralParallelFileSystem
+fs.AbstractFileSystem.gpfs.impl=org.apache.hadoop.fs.gpfs.GeneralParallelFs
+gpfs.mount.dir=/user/scratchphanpy
```
Comment (stdweird) [inline etc/hod/Hadoop-2.5.0-cdh5.3.1-gpfs/hod.conf]: `phanpy`?

File: etc/hod/Hadoop-2.5.0-cdh5.3.1-gpfs/screen.conf [hunk 2]
```diff
@@ -0,0 +1,13 @@
+#-*- cfg -*-
+# vim: ft=cfg
+
+[Unit]
+Name=screen
+RunsOn=master
+
+[Service]
+# We run 'exec bash' at the end so screen doesn't exit upon command completion.
+ExecStart=screen -dmS HODclient bash -c "module load Hadoop/2.4.0-seagate; HADOOP_CONF_DIR=$localworkdir/conf; exec bash"
```
Comment (stdweird) [inline etc/hod/Hadoop-2.5.0-cdh5.3.1-gpfs/screen.conf]: `2.4.0-seagte`? there's no varaible for this?

File: etc/hod/Hadoop-on-lustre2/screen.conf [hunk 3]
```diff
@@ -0,0 +1,14 @@
+#-*- cfg -*-
+# vim: ft=cfg
+
+[Unit]
+Name=screen
+RunsOn=master
+
+[Service]
+module=Hadoop/2.3.0-cdh5.0.0
```
Comment (stdweird) [inline etc/hod/Hadoop-on-lustre2/screen.conf]: is this correct?

**Comment by stdweird** [comment]
minor reamrks, looks ok i guess (still don't like the large amounts of identical data between all the configs)

File: etc/hod/Hadoop-2.5.0-cdh5.3.1-gpfs/screen.conf [hunk 4]
```diff
@@ -0,0 +1,14 @@
+#-*- cfg -*-
+# vim: ft=cfg
+
+[Unit]
+Name=screen
+RunsOn=master
+
+[Service]
+modules=Hadoop/2.5.0-cdh5.3.1
```
Comment (stdweird) [inline etc/hod/Hadoop-2.5.0-cdh5.3.1-gpfs/screen.conf]: `modules` or `module`? (same above too, but no templating there)

File: etc/hod/Hadoop-on-lustre2/screen.conf [hunk 5]
```diff
@@ -0,0 +1,14 @@
+#-*- cfg -*-
+# vim: ft=cfg
+
+[Unit]
+Name=screen
+RunsOn=master
+
+[Service]
+modules=Hadoop/2.4.0-seagate
```
Comment (stdweird) [inline etc/hod/Hadoop-on-lustre2/screen.conf]: same here: modules vs module

File: hod/config/config.py [hunk 6]
```diff
@@ -89,8 +89,13 @@ def _parse_runs_on(s):
     options = dict(master=RUNS_ON_MASTER, slave=RUNS_ON_SLAVE, all=RUNS_ON_ALL)
     return options[s.lower()]
 
-def _cfgget(config, section, item, dflt=None):
-    '''Get a value from a ConfigParser object or a default if it's not there.'''
+def _cfgget(config, section, item, dflt=None, **kwargs):
```
Comment (stdweird) [inline hod/config/config.py]: does this work? (named arg + the `**kwargs` magic)

File: hod/config/config.py [hunk 7]
```diff
@@ -89,8 +89,13 @@ def _parse_runs_on(s):
...
+    '''
+    Get a value from a ConfigParser object or a default if it's not there.
+    Options in kwargs come from the command line and override the config.
```
Comment (stdweird) [inline hod/config/config.py]: `from the command line` makes no sense here. just put `Options in kwargs override the config`

File: hod/config/config.py (hunk 6 above)
Comment (ehiggs) [inline hod/config/config.py]: It sure does!
```python
    In [1]: def foo(a, b, **kwargs):
       ...:     return a + b + sum([v for k,v in kwargs.items()])
       ...: 

    In [2]: foo(1, 2, x=3, **dict(y=4, z=5))
    Out[2]: 15
```

File: hod/config/config.py [hunk 8]
```diff
@@ -117,11 +122,11 @@ class PreServiceConfigOpts(object):
 
     OPTIONAL_FIELDS=['master_env', 'modules', 'service_configs', 'directories']
 
-    def __init__(self, fileobj):
+    def __init__(self, fileobj, **kwargs):
         _config = load_service_config(fileobj)
         self.version = _cfgget(_config, _META_SECTION, 'version', '')
```
Comment (stdweird) [inline hod/config/config.py]: why no `**kwargs` here too?

File: hod/hodproc.py [hunk 9]
```diff
@@ -130,7 +131,8 @@ def distribution(self, *master_template_args, **kwargs):
         m_config_filenames = parse_comma_delim_list(m_config_filenames)
 
         self.log.info('Loading "%s" manifest config', m_config_filenames)
-        m_config = preserviceconfigopts_from_file_list(m_config_filenames)
+        m_config = preserviceconfigopts_from_file_list(m_config_filenames,
+                workdir=self.options.options.config_workdir)
```
Comment (stdweird) [inline hod/hodproc.py]: move the whole block that is common with `ConfiguredMaster` in new private method.

File: hod/config/config.py (hunk 6 above)
Comment (stdweird) [inline hod/config/config.py]: ok, just looks very hackish :smile:

File: hod/hodproc.py (hunk 9 above)
Comment (ehiggs) [inline hod/hodproc.py]: Yeah I think there's a lot of repetition here but there's not really a semantic grouping I can see here. It's basically manually compressing code for the sake of the [DRY principle](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself).

I'll take a look, though.

**Pull request merged**